src/DS/metrics_extraction.py

```diff
@@ -0,0 +1,300 @@
+
+    result.update(
+        {
+            "num_test_cases": num_test_cases,
+            "num_assertions": num_assertions,
+            "has_setup": has_setup,
+            "has_teardown": has_teardown,
+            "complexity": complexity,
+        }
+    )
+
+    return result
+
+
+def analyse_code_file(file_path, max_complexity=10):
+    """
+    Analyses a Python code file to extract code metrics such as cyclomatic
+    complexity, lines of code, and the number of functions.
+
+    Args:
+        file_path (str): The path to the code file.
+        max_complexity (int): The maximum allowed complexity for a function.
+
+    Returns:
+        dict: A dictionary containing the code metrics.
+    """
+
+    logger.info(f"Starting analysis of code file: {file_path}")
+
+    result = {
+        "cyclomatic_complexity": 0,
+        "lines_of_code": 0,
+        "num_functions": 0,
+    }
+
+    parsed_file = read_and_parse_file(file_path)
+    if parsed_file is None:
+        logger.warning(f"Analysis skipped for file: {file_path}")
+        return result
+
+    tree, content = parsed_file
+
+    num_functions = 0
+    lines_of_code = len(content.splitlines())
+    cyclomatic_complexity = 0
+
+    visitor = PathGraphingAstVisitor()
+    visitor.preorder(tree, visitor)
+
+    for graph in visitor.graphs.values():
+        "Each graph in visitor.graphs.values() represents the control flow " "graph (CFG) of a single function, which is used to compute its " "cyclomatic complexity."
+        "'PathGraphingAstVisitor' processes the Abstract Syntax Tree (AST) of" "the code and generates a control flow graph for each function it" "encounters, storing these graphs in 'visitor.graphs'. Each entry" "in 'visitor.graphs' corresponds to one function, so iterating over" "it and counting with 'num_functions += 1' accurately captures the" "function count."
+
+        num_functions += 1
+        complexity = graph.complexity()
+        cyclomatic_complexity += complexity
+        logger.debug(f"Function {graph.name} has complexity {complexity}")
+
+    result.update(
+        {
+            "cyclomatic_complexity": cyclomatic_complexity,
+            "lines_of_code": lines_of_code,
+            "num_functions": num_functions,
+        }
+    )
+
+    return result
+
+
+if __name__ == "__main__":
+    args = parse_args()
+    input_file = args.input
+    output_file = args.output
+
+    working_directory = get_working_directory_or_git_root()
+    logger.info(f"Working directory: {working_directory}")
+
+    df = pd.read_csv(input_file)
+    df_language_python = df[df["guessed_language"] == "Python"]
+
+    final_df_path = output_file
+    logger.info("Apply the analysis to each test file")
+
+    # Initialise or load the final Dataframe
+    if Path(final_df_path).exists():
+        final_df = pd.read_csv(final_df_path)
+        logger.info(f"Loaded existing final Dataframe from {final_df_path}")
```

Comment:
If the dataframe already exists as a file, does this mean the results may contain duplicate entries (assuming the input file's content don't change)? Are you planning to implement a recovery mechanism, similar to what we implemented when cloning and analysing git repos? In other words, what's the value in loading previous analysis and appending results to it?

Reply:
Yes I implemented a recovery mechanism.
- Added logic to skip already-processed files by checking the existing output file   
for unique `file_path` entries.
- Improved logging to track the number of processed files and skipped files